File: stripe/src/test/java/com/stripe/android/model/AddressTest.java
package com.stripe.android.model;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static com.stripe.android.testharness.JsonTestUtils.assertMapEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

/**
 * Test class for {@link Address}.
 */
public class AddressTest {

    static final String JSON_ADDRESS = "{" +
            "\"city\": \"San Francisco\"," +
            "\"country\": \"US\",\n" +
            "\"line1\": \"123 Market St\"," +
            "\"line2\": \"#345\"," +
            "\"postal_code\": \"94107\"," +
            "\"state\": \"CA\"" +
            "}";

    private static final Map<String, Object> MAP_ADDRESS = new HashMap<String, Object>() {{
        put("city", "San Francisco");
        put("country", "US");
        put("line1", "123 Market St");
        put("line2", "#345");
        put("postal_code", "94107");
        put("state", "CA");
    }};

    private static final Address ADDRESS = Address.fromString(JSON_ADDRESS);

    @Test
    public void fromJsonString_toMap_createsExpectedMap() {
        assertNotNull(ADDRESS);
        assertMapEquals(MAP_ADDRESS, ADDRESS.toMap());
    }

    @Test
    public void builderConstructor_whenCalled_createsExpectedAddress() {
        final Address address = new Address.Builder()
                .setCity("San Francisco")
                .setCountry("US")
                .setLine1("123 Market St")
                .setLine2("#345")
                .setPostalCode("94107")
                .setState("CA")
                .build();
        assertEquals(address, ADDRESS);
    }
}

File: stripe/src/test/java/com/stripe/android/model/AddressTest.kt
package com.stripe.android.model

import com.stripe.android.testharness.JsonTestUtils.assertMapEquals
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Test class for [Address].
 */
class AddressTest {

    @Test
    fun fromJsonString_toParamMap_createsExpectedParamMap() {
        assertMapEquals(MAP_ADDRESS, ADDRESS.toParamMap())
    }

    @Test
    fun builderConstructor_whenCalled_createsExpectedAddress() {
        val address = Address.Builder()
            .setCity("San Francisco")
            .setCountry("US")
            .setLine1("123 Market St")
            .setLine2("#345")
            .setPostalCode("94107")
            .setState("CA")
            .build()
        assertEquals(address, ADDRESS)
    }

    @Test
    fun toParamMap_shouldRejectEmptyValues() {
        val addressParams = Address.Builder()
            .setCountry("US")
            .build()
            .toParamMap()
        assertEquals(mapOf("country" to "US"), addressParams)
    }

    companion object {
        private val MAP_ADDRESS = mapOf(
            "city" to "San Francisco",
            "country" to "US",
            "line1" to "123 Market St",
            "line2" to "#345",
            "postal_code" to "94107",
            "state" to "CA"
        )

        private val ADDRESS = Address.fromJson(AddressFixtures.ADDRESS_JSON)!!
    }
}
